File: contents/02/basic-2/solution.py
class BinaryHeap:

    def __init__(self, num_node):
        n = 1
        self.vacant_idx = 0
        while n < num_node:
            n *= 2
        self.dat = [0]*(n*2)

    def push(self, x, dat=None):
        if dat == None:
            dat = self.dat
        dat[self.vacant_idx] = x
        cur = self.vacant_idx
        self.vacant_idx += 1
        par = (cur-1)//2
        while cur > 0 and dat[par] < dat[cur]:
            dat[par], dat[cur] = dat[cur], dat[par]
            cur = par
            par = (cur-1)//2

    def pop(self, dat=None):
        if dat == None:
            dat = self.dat
        ret = dat[0]
        self.vacant_idx -= 1
        dat[0], dat[self.vacant_idx] = dat[self.vacant_idx], 0
        cur = 0
        while True:
            vl = dat[cur*2+1]
            vr = dat[cur*2+2]
            if vl > vr and vl > dat[cur]:
                dat[cur], dat[cur*2+1] = dat[cur*2+1], dat[cur]
                cur = cur*2+1
            elif vr > dat[cur]:
                dat[cur], dat[cur*2+2] = dat[cur*2+2], dat[cur]
                cur = cur*2+2
            else:
                break
        return ret


Q = int(input())
heap = BinaryHeap(Q)
for _ in range(Q):
    lis = list(map(int, input().split()))
    if lis[0] == 1:
        heap.push(lis[1])
    else:
        res = heap.pop()
        print(res)

File: contents/02/basic-2/solution.cpp
#include<bits/stdc++.h>

using namespace std;

template<typename T>
struct BinaryHeap {
	int n;
	int r;
	vector<T>dat;
	BinaryHeap(const int Q) {
		n = 1;
		r = 0;
		while (n<Q)n *= 2;
		dat.assign(n * 2, 0);
	}
	void push(T x) {
		dat[r] = x; 
		int pos = r;
		r++;
		while (pos > 0 && dat[(pos - 1) / 2] < dat[pos]) {
			swap(dat[(pos - 1) / 2], dat[pos]);
			pos = (pos - 1) / 2;
		}
	}
	int pop() {
		int ret = dat[0];
		dat[0] = dat[r-1]; 
		dat[r-1] = 0; r--;
		int pos = 0;
		while (true) {
			int vl = dat[pos * 2 + 1];
			int vr = dat[pos * 2 + 2];
			if (vl > vr) {
				if (vl > dat[pos]) {
					swap(dat[pos], dat[pos * 2 + 1]);
					pos = pos*2+1;
				}
				else break;
			}
			else {
				if (vr > dat[pos]) {
					swap(dat[pos], dat[pos * 2 + 2]);
					pos = pos*2+2;
				}
				else break;
			}
		}
		return ret;
	}
};

void solve() {
	int Q;cin>>Q;
	BinaryHeap<int>heap(Q);
	while(Q--){
		int sign;cin>>sign;
		if(sign==1){
			int x;cin>>x;
			heap.push(x);
		}
		else{
			cout<<heap.pop()<<endl;
		}
	}
}

signed main() {
	solve();
}
